src/D2L.CodeStyle.Analyzers/ApiUsage/ImmutableGenericAttributeAnalyzer.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class ImmutableGenericAttributeAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+			=> ImmutableArray.Create(
+				Diagnostics.ImmutableGenericAttributeInWrongAssembly,
+				Diagnostics.ImmutableGenericAttributeAppliedToNonGenericType,
+				Diagnostics.ImmutableGenericAttributeAppliedToOpenGenericType
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.RegisterCompilationAction( AnalyzeAssembly );

```

Comment:
Would concurrent execution be possible?

Reply:
Yeah, but wouldn't help. We really only want 1 symbol (the assembly symbol), so that we can get all attributes applied to it. We can analyze the attributes in parallel, but sadly there is no `IAttributeSymbol` (logically), so we'd have to do the parallelism ourselves.

Reply:
Oh right, there is only one assembly to analyze :P I get it